### PR TITLE
WIP: entry_verification_service

### PR DESCRIPTION
**Pull request by ilya-bobyr**
#### Problem

Currently, PoH and signature verification for entries is only performed when all the preceding entries are processed.  As entries could be received out of order, it is possible for of the verification work to be pending.  Waiting for the predecessor data set to be processed.

#### Summary of Changes

Introduce a verification service, that runs PoH and transaction signature verification out of order.  Making as much progress as possible based on the available data.

Fixes [#28633](https://github.com/solana-labs/solana/issues/28633)

### CODE REVIEW

File: core/src/entry_verification_service/shared_wrapper.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
+//! A simple wrapper for state shared between threads, that can also be waited on.
```
Comment (bw-solana) [inline core/src/entry_verification_service/shared_wrapper.rs:1]: This feels like it could be its own standalone crate. I don't see anything in here very specific to the entry verification service

File: core/src/entry_verification_service.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        slot: Slot,
+        data_set: DataSetFirstShred,
+    ) -> Result<bool, WaitForResultError> {
+        self.commands
+            .send(Command::WaitingFor { slot, data_set })
+            .map_err(|_| WaitForResultError::WorkerTerminated)?;
+
+        let wait_start = Instant::now();
+        match self.results.check_and_wait_until(deadline, |results| {
+            if slot < results.starting_at {
```
Comment (bw-solana) [inline core/src/entry_verification_service.rs:74]: Could use a comment here explaining what this means. I'm assuming we're waiting on a slot that has already been processed or something

File: core/src/entry_verification_service.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        Self {
+            starting_at: 0,
+            slots: HashMap::new(),
+        }
+    }
+}
+
+impl Results {
+    fn set_slot_status(&mut self, slot: Slot, status: SlotVerificationStatus) {
+        if slot < self.starting_at {
```
Comment (bw-solana) [inline core/src/entry_verification_service.rs:151]: Is this expected or an error condition?

File: core/src/entry_verification_service/worker.rs [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+        entries: Arc<Vec<Entry>>,
+    },
+    /// Indicates that someone is waiting for a validation result for this slot.  This is just a
+    /// hint to start validation ASAP.
+    WaitingFor {
+        slot: Slot,
+        data_set: DataSetFirstShred,
+    },
+    /// Remove state for slots before the specified one.
+    Cleanup { earlier_than: Slot },
```
Comment (bw-solana) [inline core/src/entry_verification_service/worker.rs:37]: Who is going to be responsible for issuing Cleanup? I'm thinking of the ways one thread could cleanup state that another thread was waiting on

File: core/src/entry_verification_service/worker.rs [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+        ///
+        /// Entries are delivered in chunks, called "data sets".  Each data set is contained in
+        /// a consecutive range of shreds, that were used to deliver those entries.
+        ///
+        /// This map only contains entries at or after `verified_up_to`.
+        data_sets: HashMap<DataSetFirstShred, DataSetOptimisticStatus>,
+    },
+
+    /// All entries in the slot have been successfully verified.  We still need to remember the very
+    /// last entry hash, as it is needed for verification of any slots that use this one as a
```
Comment (bw-solana) [inline core/src/entry_verification_service/worker.rs:92]: *needed for verification of any children of this slot

File: core/src/entry_verification_service/worker.rs [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+            let todo = match self.accumulate() {
+                Some(todo) => todo,
+                None => return,
+            };
+
+            self.verify(todo);
+        }
+    }
+
+    fn accumulate(&mut self) -> Option<Todo> {
```
Comment (bw-solana) [inline core/src/entry_verification_service/worker.rs:267]: maybe `accumulate_verification_requests` would be more explicit

Comment (bw-solana) [inline core/src/entry_verification_service/worker.rs:267]: Not sure how best to do this, but it would be great to add some metrics so we can gauge the effectiveness of accumulate.

I'm assuming our hope is that by waiting we can avoid the over head of, for example, verifying a dataset except for the first element, then the preceding entry comes in and we loop back to verify the first element (as opposed to just waiting a little bit and then verifying everything in one shot).

File: core/src/entry_verification_service/worker.rs [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+    },
+};
+
+mod commands;
+
+pub(super) enum Command {
+    /// Add new data set for the specified slot.
+    AddDataSet {
+        slot: Slot,
+        shard_indices: RangeInclusive<u32>,
```
Comment (bw-solana) [inline core/src/entry_verification_service/worker.rs:26]: shred* ?

File: core/src/entry_verification_service/worker.rs [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+
+        // TODO Run signature verification.
+
+        let failed = vec![];
+        // TODO Check results and mark any failed slots.
+
+        for slot in failed.iter() {
+            if let Some(i) = verified.iter().position(|v| *v == *slot) {
+                verified.swap_remove(i);
+            }
```
Comment (bw-solana) [inline core/src/entry_verification_service/worker.rs:386]: Would None here be considered an error?

File: core/src/entry_verification_service/shared_wrapper.rs (hunk 1 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service/shared_wrapper.rs:1]: True.  But I didn't want to generalize too early.
So far, I'm writing what is needed for this service.
I didn't see this pattern in the code that I've seen so far, so I wrote it.
If another service will needed it, then it can be adjusted and moved into a shared location.
My personal experience is that it makes sense to write 3 specific versions, and only then to generalize them.  As, often, a generalization that comes from a single case is far from ideal.
But if you think I should put it into a shared location, just let me know.

File: core/src/entry_verification_service.rs (hunk 2 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service.rs:74]: There is an explanation in the doc-comment for `Result`:
```rust
/// Results that are shared between the worker and the API of the service.
/// Wrapping it in [`SharedWrapper`] allows the API to wait for the worker, without actively
/// checking for the state to be updated.
struct Results {
    /// First slot that we are going to process and, possibly, have any results for.
    ///
    /// `EntryVerificationService::cleanup()` shifts this value forward.  And all the other
    /// operations are only valid, when called for this or later slots.
    starting_at: Slot,

    slots: HashMap<Slot, SlotVerificationStatus>,
}
```

I try to document the data structures first.
It is accessible in the LSP hints, though, you do not get that in GitHub %)
I think `Results::starting_at` doc-comment explains it.
But let me know if you think this deserves an additional explanation.

For the `set_slot_status()` case, ignoring an update for a slot that is before `starging_at` is the only choice.  As the data structure invariant is that we should not have any information in `slots`, for slots before `starting_at`.

Returning an error is an option, in general.  But in this particular case, it does not matter to any of the callers.  As it would be just a race with the cleanup, if everything else is working correctly.  I do not think I can avoid this race.  Though, I have not though about it too much.

File: core/src/entry_verification_service.rs (hunk 3 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service.rs:151]: Addressed this in the thread above.
Let's continue there.

File: core/src/entry_verification_service/worker.rs (hunk 4 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service/worker.rs:37]: This is a connection point to the rest of the system.
`Cleanup` has to be called for a slot that can not be used as a parent of a fork anymore.
I was trying to make sure that the service stores as little information as possible for each slot.  So, storing data on a hundred or a thousand slots, that have been already processed, should not be expensive.
But still, for a long-running validator there should be a regular cleanup.

If it so happens that there is someone waiting for validation results on a slot that is marked as too old by `Cleanup`, they would just receive a wake-up notification and the slot will be considered failed.

This happens due to the fact that `Results` can only be updated via a `SharedWrapper::update_notify_all()` call.  And anybody reading the state, needs to do so via a `SharedWrapper::check_and_wait_until()` call.

`EntryVerificatonService::wait_for_result()` does this:
```rust
    pub fn wait_for_result(
        &self,
        deadline: Instant,
        slot: Slot,
        data_set: DataSetFirstShred,
    ) -> Result<bool, WaitForResultError> {
        /* ... */
        match self.results.check_and_wait_until(deadline, |results| {
            if slot < results.starting_at {
                return Some(false);
            }
            /* ... */
        }) {
            Some(valid) => Ok(valid),
            /* ... */
        }
    }
```

So if `slot < results.starting_at`, the return value is `false`.
Either immediately, or if `results` is updated in parallel.

File: core/src/entry_verification_service/worker.rs (hunk 5 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service/worker.rs:92]: Applied.

File: core/src/entry_verification_service/worker.rs (hunk 6 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service/worker.rs:267]: This is a local function that is only called by `run()`.
For such a small scope, it seems nice to use short names.
`run()` is just
```rust
        loop {
            let todo = match self.accumulate() {
                Some(todo) => todo,
                None => return,
            };

            self.verify(todo);
        }
```

Everything is about verification here, as this is a verification service :)
Maybe `accumulate_work()` could be a clarification?
You are right that `accumulate()` by itself could refer to many different things %)
What do you think?

Comment (ilya-bobyr) [reply core/src/entry_verification_service/worker.rs:267]: Absolutely.
I did not write the metrics code yet, but there are a number of things that would be good to measure.
In particular, I wanted to measure the lengths of the `commands` channel.
As it is an indication of how useful the accumulation really is.

As for the wait.  I think, it could actually work forwards and across forks too.
Like we could wait a bit, and possibly, get the next data set, if there is nobody actively waiting for the current data set.
In which case, we can run verification for two data sets at a time.
This could happen if we missed a data set, which would make the replay service wait for a retransmission or a repair.
But the verification service can continue running for subsequent data sets.
It will still run both PoH and signature verification.  It will just skip the very first entry for the PoH hashing.

If not under pressure from the replay service, there might be savings in combining multiple requests together.  From what I read in the verification code (both PoH and signature), there are optimizations to run it in parallel, when enough work is available.
So, I figured, this way, we may also combine completely unrelated data sets together.
For example, when two forks are active, this service will be just fine doing a single GPU invocation for data from both forks.  When they happen to send data sets at approximately the same time.

I also have no idea what would be a good accumulation delay.
Replay service will send `WaitingFor` command when it is actually waiting for something.
So the accumulation delay should not affect the replay service performance [1].
And the only question is: is it accumulating enough to use the efficiency of the parallel verification?
I will need to add metrics to measure the `self.verify` invocation delay.
And the `todo` length.
In addition to the `commands` channel length.
I think measuring averages for these parameters could provide a good understanding if this mechanism is working or not.

[1] I've added a discussion on the `WaitingFor` command potential delays to the enum entry that defines the command:

```rust
    /// There is a chance for this command delivery to be delayed.  If the `commands` channel has
    /// some unfinished work and `todo` has a lot of accumulated work, then we might not start the
    /// verification work for things that the replay service might need us to verify right now.
    ///
    /// Ideally, this should not happen during normal operation.  And should only happen during very
    /// high load, when there is not enough CPU to process all the requests.  In which case, it
    /// seems fine.  Accumulation delay might trigger processing then.
    ///
    /// We need to add metrics to make sure that neither the `commands` channel, nor the `todo` list
    /// do grow too long under normal operation.
    ///
    /// An alternative design would be to send `WaitingFor` though a separate channel.  But it seems
    /// that if it really matters, then the verification is not working fast enough in the first
    /// place.  Also we do not currently track which pending work is needed for `slot`/`data_set` to
    /// be verified, and which work might potentially be delayed for later.  So it is not as simple
    /// as just adding another channel.
```

File: core/src/entry_verification_service/worker.rs (hunk 7 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service/worker.rs:26]: Fixed.

Seems like I'm not the only one:
```
❯ git grep -w shard | wc -l
53

❯ git grep -w shard ledger/src/shred | wc -l
39
```

File: core/src/entry_verification_service/worker.rs (hunk 8 above)
Comment (ilya-bobyr) [reply core/src/entry_verification_service/worker.rs:386]: It would.  But I was trying to avoid `unwrap()` and `panic!()` calls.
`unexpected_error!` will just report an error in a release build and continue.
Here, `None` should never happen, yes.  It would be a logic error in the `verify()` call logic.
I can change `if` to `expect()`.  But it does not seem to be that bad that we want to stop the validator.  Not sure.
It should also be checked by unit tests anyway.  I will write some for this method when it will be connected to an actual verification call.

File: ledger/src/blockstore.rs [hunk 9]
```diff
@@ -3379,29 +3386,61 @@ fn update_completed_data_indexes(
...
-    shred_indices
-        .windows(2)
-        .filter(|ix| {
-            let (begin, end) = (ix[0] as u64, ix[1] as u64);
-            let num_shreds = (end - begin) as usize;
-            received_data_shreds.range(begin..end).count() == num_shreds
-        })
-        .map(|ix| (ix[0], ix[1] - 1))
-        .collect()
 }
```
Comment (behzadnouri) [inline ledger/src/blockstore.rs:3444]: I think we had discussed this change previously on a different PR.
Is this change necessary and needs to be included with this PR?
The change is already pretty big.

**Comment by behzadnouri** [comment]
* On a cursory look this change (3.5k loc so far) is pretty large for what it describes to make. Should we discuss first the design and understand why such a small functional change has become so large in terms of amount of code? 
* Given the added complexity, do we want to commit to going down this route without measuring first if this is the appropriate path? I mentioned [some concerns on the tracking issue](https://github.com/solana-labs/solana/issues/28633#issuecomment-1412681761) that it is not given that the trade-offs are positive and we probably need to confirm that first.

File: ledger/src/blockstore.rs (hunk 9 above)
Comment (ilya-bobyr) [reply ledger/src/blockstore.rs:3444]: No, it is not necessary.
It got in here by accident - it was still in my `master`, that I published in this PR.
The only interesting change is the [very last commit on the branch](https://github.com/solana-labs/solana/pull/30488/commits/b52b64dcf04f65b79f0d32fa67d5d6345c11f42b).
The other commits that preceded the last one are necessary, though.

**Comment by ilya-bobyr** [comment]
> * On a cursory look this change (3.5k loc so far) is pretty large for what it describes to make. Should we discuss first the design and understand why such a small functional change has become so large in terms of amount of code?

A lot of it is tests and comments.
Though, some of it is changes in functionality, yes.
How do you want to discuss the design?

I was planning to add a general description of the implementation design to the service doc-comments.
But as I was trying to get to the point where I can do some measurements first, it was lower on my todo list.

> * Given the added complexity, do we want to commit to going down this route without measuring first if this is the appropriate path? I mentioned [some concerns on the tracking issue](https://github.com/solana-labs/solana/issues/28633#issuecomment-1412681761) that it is not given that the trade-offs are positive and we probably need to confirm that first.

Yes, I remember that.
And I am trying to figure out a way to test it early.
At the same time, for example, without the unit tests, I would probably spend more time trying to find bugs when the whole process would fail at runtime.
I am still relatively new to the codebase.  So it is possible that I am missing something obvious.
But it is not completely clear to me how exactly can it be measured without actually having working code.

I think it is possible to measure the baseline.
By removing the validation altogether, and running a validator with some additional metrics.
But it would not really show if this particular approach is better or not.

**Comment by ilya-bobyr** [comment]
Here is a high level overview of the design, if you will be looking into this more.
I'll copy it into doc-comments later.

---

Both PoH and signature verification for entries is done by a separate verification service.
It has a relatively simple API:
1. [Add data set](https://github.com/solana-labs/solana/pull/30488/files#diff-402025f43304978f58e996d343a3da641e4eef5473395288e0cc93e1f7c29921R45) to the verification process.
2. [Check and wait](https://github.com/solana-labs/solana/pull/30488/files#diff-402025f43304978f58e996d343a3da641e4eef5473395288e0cc93e1f7c29921R62) if a particular slot is valid up to a given data set.
3. [Discard state](https://github.com/solana-labs/solana/pull/30488/files#diff-402025f43304978f58e996d343a3da641e4eef5473395288e0cc93e1f7c29921R99) for all slots earlier than a given one.

Window service calls `add_data_set()`.
Replay service calls `wait_for_result()`.
I still need to figure out how to connect the `cleanup()` call.
Brennan was suggesting it can happen when a root is formed.

Internally, the service remembers the data sets it received, and runs PoH and signature verification for everything it got.
When entries form a contiguous chain, it updates the verification results for corresponding slots.

Most of the complexity comes from two sources:

1. Data sets can be received in arbitrary order.
   Yet, for PoH hashes, one needs the preceding data set to check the next.
   As data sets contain multiple entries, if the preceding data set is absent, it is still possible to run PoH verification for all by the first entry in the data set.
   Later, when the starting hash for the first entry becomes available, the first entry can be verified.
   There is an additional wrinkle, when the data set in question is the first in slot.  As in this case, we need to check the parent slot for the right hash.  And as slots for a tree, it is not entirely trivial, especially when some data might be missing.

2. As there is a considerable number of data sets, I was trying to optimize all operations to have the minimum number of lookup and memory accesses.  And to use as less memory as possible, discarding everything that is no longer unnecessary.

File: core/src/entry_verification_service/worker.rs (hunk 6 above)
Comment (bw-solana) [reply core/src/entry_verification_service/worker.rs:287]: `accumulate_work()` works too. We use accumulate a lot of places for metrics, so I'll be happy with any postfix to save my future self :)

Comment (ilya-bobyr) [reply core/src/entry_verification_service/worker.rs:287]: Done.

**Comment by ilya-bobyr** [comment]
This is still WIP, but here is a summary of the latest update:
1. Removed unrelated commits from the PR.  Sorry about those.
2. Completed test coverage.
   It uncovered a minor improvement in the data structure design, and a few minor issues.
3. Added code for both PoH and signature verification.  No GPU or SIMD support.
4. Address most of the comments.  Except for the metrics.
  4.1. Added a general description to the service.  Copied from the PR messages.
  
Next, I'll connect the service to the replay stage to start measuring performance.
